### DISCUSSION TITLE
Widget with @Work: how to self remove?

### DISCUSSION THREAD

**Discussion by jehon** [Textualize/textual#2837] [Q&A] [answered]
I have a widget that display the logs of a command. Upon completion, the widget should self-destruct:

```
class ExecLogWidget(TextLog):
    def __init__(self):
        self.start()
    @work(exclusive=True)
    def start(self) -> None:
            with subprocess.Popen(self.command, stdout=subprocess.PIPE, text=True, universal_newlines=True) as proc:
                if proc.stdout is not None:
                    for stdout_line in proc.stdout:
                        self.write(stdout_line)
            self.write("")
            self.write("[green] ✓ Success")
            time.sleep(5)
            self.remove()

```
It seems that the `self.remove()` does not work, because it is in a separated thread, and break the cycle of the main thread:

```
RuntimeError: no running event loop
sys:1: RuntimeWarning: coroutine 'App._remove_nodes.<locals>.prune_widgets_task' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

I should use "call_from_thread", but I don't have access to the "App" object from inside this widget.

Any idea?

**Comment by willmcgugan** [comment, marked as answer]
You should be able to get a reference to the app with `self.app`.

Note that you probably shouldn't call `self.start` from the initializer. Better to call it from `on_mount`, to ensure the app and widget are running.

**Reply by jehon** [reply]
It worked, thanks a lot!
